### DISCUSSION TITLE
sqlalchemy.exc.ArgumentError: Mapper mapped class User->users could not assemble any primary key columns for mapped table 'users'

### DISCUSSION THREAD

**Discussion by Timtaran** [sqlalchemy/sqlalchemy#5771] [Usage Questions] [answered]
**Describe the bug**

**Expected behavior**

**To Reproduce**
Please try to provide a [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) example.
See also [Reporting Bugs](https://www.sqlalchemy.org/participate.html#bugs) on the website, and some [example issues](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22).

https://pastebin.com/GF3abSKd

**Error**

```
Traceback (most recent call last):
  File "C:/Users/user/Documents/GitHub/pythonProject/LifeSimulator/src/tests/test1.py", line 11, in <module>
    class User(Base):
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\api.py", line 76, in __init__
    _as_declarative(cls, classname, cls.__dict__)
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\base.py", line 131, in _as_declarative
    _MapperConfig.setup_mapping(cls, classname, dict_)
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\base.py", line 160, in setup_mapping
    cfg_cls(cls_, classname, dict_)
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\base.py", line 194, in __init__
    self._early_mapping()
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\base.py", line 199, in _early_mapping
    self.map()
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\ext\declarative\base.py", line 695, in map
    self.cls.__mapper__ = mp_ = mapper_cls(
  File "<string>", line 2, in mapper
  File "<string>", line 2, in __init__
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\util\deprecations.py", line 139, in warned
    return fn(*args, **kwargs)
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\orm\mapper.py", line 723, in __init__
    self._configure_pks()
  File "C:\Users\user\Documents\GitHub\pythonProject\LifeSimulator\venv\lib\site-packages\sqlalchemy\orm\mapper.py", line 1409, in _configure_pks
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Mapper mapped class User->users could not assemble any primary key columns for mapped table 'users'
```

**Versions.**
 - OS: Windows 7 64-bit
 - Python: 3.8.6
 - SQLAlchemy: 1.3.20
 - Database: sqlite3
 - DBAPI:

**Additional context**

**Have a nice day!**

**Comment by CaselIT** [comment, marked as answer]
Hi,

As the exception message says, sqlalchemy "could not assemble any primary key columns for mapped table 'users'". The sqlalchemy orm needs a primary key for each mapped table.
See https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.Column.params.primary_key for how to set a primary key on a column

**Reply by naeemnawaz** [reply]
salam, everyone!
in every table/class it is necessary to put primary_key= True on one column

**Reply by CaselIT** [reply]
One or more. The orm supports composite primary keys

**Reply by micheldeslandes** [reply]
If you arrived here because of an error while copying, line by line, an example given on the docs.sqlalchemy.org page, you probably overlooked the helpful "copy button" and a Python indentation problem. The empty line needs indentation.

>>> from sqlalchemy import Column, Integer, String
>>> class User(Base):
...     __tablename__ = 'users'
...      <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< Although empty, this line needs to be indented. 
...     id = Column(Integer, primary_key=True)
...     name = Column(String)
...     fullname = Column(String)
...     nickname = Column(String)
...
...     def __repr__(self):
...        return "<User(name='%s', fullname='%s', nickname='%s')>" % (
...                             self.name, self.fullname, self.nickname)


EDIT: Or you can just click on the helpful "copy" button that appears on the top-right of the light grey example box, when you hover your mouse over it, and then paste it onto the Python prompt or into your script. That way the newline will *not* be copied.

**Reply by CaselIT** [reply]
@micheldeslandes could you link to the documentation page that has that issue so that it can be fixed?

**Reply by micheldeslandes** [reply]
https://docs.sqlalchemy.org/en/14/orm/tutorial.html#declare-a-mapping

**Reply by zzzeek** [reply]
@micheldeslandes  cant reproduce, I clicked the "copy" box in the example and formatting is correct, blank lines are omitted so their formatting doesn't matter:

```py
from sqlalchemy import Column, Integer, String
class User(Base):
    __tablename__ = 'users'
    id = Column(Integer, primary_key=True)
    name = Column(String)
    fullname = Column(String)
    nickname = Column(String)
    def __repr__(self):
       return "<User(name='%s', fullname='%s', nickname='%s')>" % (
                            self.name, self.fullname, self.nickname)
```

**Reply by micheldeslandes** [reply]
You need to think more like a total "newby"! :-)    (like me!)

If you click in the "copy" box, it will work, because that won't copy the empty lines that are displayed in the corresponding example.

If you are oblivious to that helpful feature and just copy and paste with your mouse, line by line. And add a blank line (with no indentation) after "___tablename___" definition, you will get the error.

NOTE: I've updated my original comment to reflect the usefulness of the "copy" box.

**Reply by zzzeek** [reply]
you can't copy and paste with your mouse because it copies all the ellipsis as well, and these are not part of the Python code.

**unless**, you happen to be using iPython notebook, which has a special feature that normalizes for the ellipsis.  That's pretty much the only way direct copy-paste is possible.

**Reply by micheldeslandes** [reply]
You underestimate my ignorance, utter newbyness and total lack of professionalism!

I copy & pasted onto a notepad and then edited out the extra bits. Then pasted line-by-line, to see the system react. My suspicion is, the original poster and a few other people who will Google that same error message and stumble on this page, will do the same. That's why I mentioned this here. The documentation and the examples are fine. It's a lack of understanding on the part of the user, problem.

In my defence, though. I'm not developing and, therefore, not using an IDE. I was just testing out a few examples to see how sqlalchemy works, so I can make some scripts to perform some sanity checks on a database that I have.

If it serves as any redemption: 1) I don't work as a software developer. and 2) I would use vscode, if developping.

**Reply by Shashank1219** [reply]
Hi @CaselIT, I am using SQL Server on the backend, and I have used the IDENTITY function to produce a new id every time. How do I handle this in my Flask app? I do not want to enter new values to override the SQL as it would also give me an error. Could you please let me know how to handle this? Any help is much appreciated.
![image](https://user-images.githubusercontent.com/31164761/229903842-f1b1b220-76f0-4922-ae64-4cc9e1b8fd34.png)

**Reply by CaselIT** [reply]
Not sure why this was posted here, since it does not seem related. Please open another discussion with the appropriate information, thanks

**Comment by naeemnawaz** [comment]
I'm sorry man I couldn't work on SQL server I have not much experience in it
